File: core/tennis_names.py
"""Shared tennis player name cleanup.

Feeds disagree on punctuation and sometimes include seeds, countries or score
suffixes. Keep display names readable, but use canonical keys for lookups.
"""
from __future__ import annotations

import re
import unicodedata

_SEEDING = re.compile(r"\(\d+\)\s*")
_NATION = re.compile(r"\([A-Z]{2,3}\)")
_SCORE_SUFFIX = re.compile(
    r"\s+\d+(?:[-–]\d+)?(?:\(\d+\))?(?:\s+\d+(?:[-–]\d+)?(?:\(\d+\))?)*(?:\s+(?:ret|w/o|wo|walkover))?\s*$",
    re.IGNORECASE,
)
_PUNCT = re.compile(r"[^a-z0-9\s]")


def clean_player_name(raw: str | None) -> str:
    """Return a display-safe player name without provider noise."""
    if not raw:
        return ""
    name = str(raw)
    name = _SEEDING.sub("", name)
    name = _NATION.sub("", name)
    name = _SCORE_SUFFIX.sub("", name)
    name = re.sub(r"\s+", " ", name)
    return name.strip(" -–")


def canonical_player_key(raw: str | None) -> str:
    """Return a provider-insensitive key for Elo/ranking lookup."""
    name = clean_player_name(raw)
    name = unicodedata.normalize("NFKD", name).encode("ascii", "ignore").decode("ascii")
    name = name.lower().replace("-", " ")
    name = _PUNCT.sub(" ", name)
    return re.sub(r"\s+", " ", name).strip()

File: lib/tennis-names.ts
// Port TS di core/tennis_names.py (#FORTUNEPLAY-LIVE-ODDS-1). I feed divergono su
// punteggiatura, seed, nazione e score-suffix: chiavi canoniche per il lookup.
const SEEDING = /\(\d+\)\s*/g;
const NATION = /\([A-Z]{2,3}\)/g;
const SCORE_SUFFIX =
  /\s+\d+(?:[-–]\d+)?(?:\(\d+\))?(?:\s+\d+(?:[-–]\d+)?(?:\(\d+\))?)*(?:\s+(?:ret|w\/o|wo|walkover))?\s*$/i;
// Dopo toLowerCase() — rimuove char non-ASCII dopo decomposizione (equiv a encode('ascii','ignore') Python)
const PUNCT = /[^a-z0-9\s]/g;

export function cleanPlayerName(raw: string | null): string {
  if (!raw) return "";
  let name = String(raw);
  name = name.replace(SEEDING, "");
  name = name.replace(NATION, "");
  name = name.replace(SCORE_SUFFIX, "");
  name = name.replace(/\s+/g, " ");
  return name.replace(/^[\s\-–]+|[\s\-–]+$/g, "");
}

export function canonicalPlayerKey(raw: string | null): string {
  let name = cleanPlayerName(raw);
  name = name.normalize("NFKD").replace(/[^\x00-\x7F]/g, "");
  name = name.toLowerCase().replace(/-/g, " ");
  name = name.replace(PUNCT, " ");
  return name.replace(/\s+/g, " ").trim();
}
